production/terraform/modules/s3/main.tf

```diff
@@ -0,0 +1,72 @@
+provider "aws" {
+  region = var.region
+}
+
+data "aws_caller_identity" "current" {}
+
+data "aws_iam_policy_document" "oidc" {
+  statement {
+    actions = ["sts:AssumeRoleWithWebIdentity"]
+    effect  = "Allow"
+
+    condition {
+      test     = "StringEquals"
+      variable = "${var.oidc_id}:sub"
+      values   = ["system:serviceaccount:${var.namespace}:${var.serviceaccount}"]
+    }
+
+    condition {
+      test     = "StringEquals"
+      variable = "${var.oidc_id}:aud"
+      values   = ["sts.amazonaws.com"]
+    }
+
+    principals {
+      identifiers = [
+        "arn:aws:iam::${data.aws_caller_identity.current.account_id}:oidc-provider/${var.oidc_id}"
+      ]
+      type        = "Federated"
+    }
+  }
+}
+
+resource "aws_s3_bucket" "loki-data" {
+  bucket = "${var.bucket_name}"
+}
+
+resource "aws_iam_role" "loki" {
+  name               = "LokiStorage-${var.cluster_name}"
+  assume_role_policy = data.aws_iam_policy_document.oidc.json
```

Comment:
This should be done only when the caller intends to use S3 from a Loki installation on an EKS cluster.